### PR TITLE
Decouple CMS and OPL emulations

### PR DESCRIPTION
**Pull request by japsmits**
# Description

Decouple CMS and OPL emulations.

For SoundBlaster 1, 1.5 and 2, the CMS is (optionally) built-in. For these SoundBlasters, CMS can now be enabled or disabled. GameBlaster is an implementation of CMS and always enables it.

For backward compatibility, CMS can still be selected as the OPL-mode.

## Related issues

#2136
https://github.com/dosbox-staging/dosbox-staging/issues/2111

# Manual testing

Attached the debugger and used the Config commando to cycle through all possibilities.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [ ] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [ ] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by japsmits** [comment]
For some reason, "Windows MSYS2 builds / Clang x86_64 +tests" fails, but that seems to be unrelated to this PR. Marking this one as ready.

**Comment by johnnovak** [comment]
Added https://github.com/dosbox-staging/dosbox-staging/issues/2111 to the related issues section with your permission @japsmits.

**Comment by johnnovak** [comment]
> For some reason, "Windows MSYS2 builds / Clang x86_64 +tests" fails, but that seems to be unrelated to this PR. Marking this one as ready for review.

Yeah, don't worry about that one. That's some hackery I need to do to get the 0.81.1 patch out (disable MVSC & enable MSYS2, but MSYS2 is broken on main).

File: src/dosbox.cpp [hunk 1]
```diff
@@ -821,12 +821,16 @@ void DOSBOX_Init()
 	pint->Set_help("The OPL waveform is now sampled at the mixer's playback rate to avoid\n"
 	               "resampling.");
 
+	pbool = secprop->Add_bool("cms", when_idle, true);
+	pbool->Set_help("Enable or disable the CMS for SoundBlaster 1, 1.5 and 2.");
```
Comment (johnnovak) [inline src/dosbox.cpp:825]: For consistency with the other settings, let's go with this (e.g., look up `gus`, `imfc`, `ps1audio`, etc.).

`Enable CMS emulation for SoundBlaster 1, 1.5 and 2 (enabled by default).`

File: src/dosbox.cpp [hunk 2]
```diff
@@ -821,12 +821,16 @@ void DOSBOX_Init()
...
+
 	const char* oplmodes[] = {"auto", "cms", "opl2", "dualopl2", "opl3", "opl3gold", "none", nullptr};
 	pstring = secprop->Add_string("oplmode", when_idle, "auto");
 	pstring->Set_values(oplmodes);
 	pstring->Set_help("Type of OPL emulation ('auto' by default).\n"
 	                  "On 'auto' the mode is determined by 'sbtype'.\n"
-	                  "All OPL modes are AdLib-compatible, except for 'cms'.");
+	                  "All OPL modes are AdLib-compatible, except for 'cms'.\n"
+	                  "The 'cms'-type is deprecated.");
```
Comment (johnnovak) [inline src/dosbox.cpp:833]: Please remove the text about `cms` and the deprecation warning (we don't call deprecations out explicitly in the config descriptions). So only this should remain:

`All OPL modes are AdLib-compatible.`

File: src/hardware/sblaster.cpp [hunk 3]
```diff
@@ -2180,13 +2180,38 @@ OplMode find_oplmode()
...
 		}
 	}
 	return opl_mode;
 }
 
+bool find_cms() {
+	const auto sect = static_cast<Section_prop*>(control->GetSection("sblaster"));
+	assert(sect);
+
+    // Backward compatibility with existing configurations
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2194]: Formatting.

File: src/hardware/sblaster.cpp [hunk 4]
```diff
@@ -2180,13 +2180,38 @@ OplMode find_oplmode()
...
+	if (sect->Get_string("oplmode") == "cms") {
+		LOG_WARNING("The use of oplmode 'cms' is now deprecated.");
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2196]: Prefix the warning with the SB model like all the other log statements in this file and suggest a resolution. E.g.,

```cpp
LOG_WARNING("%s: The 'cms' setting for 'oplmode' is deprecated; please use `cms = on` instead.", CardType());
```

File: src/hardware/sblaster.cpp [hunk 5]
```diff
@@ -2180,13 +2180,38 @@ OplMode find_oplmode()
...
+		return true;
+	}
+   
+    auto get_cms = [sect]() {
+		auto* opt_cms = sect->GetBoolProp("cms");
+		return opt_cms == nullptr ? true
+		                          : static_cast<bool>(opt_cms->GetValue());
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2203]: Just use `sect->Get_bool("cms")`. This would be the first use of calling `GetBoolProp` in the code 😄

File: src/hardware/sblaster.cpp [hunk 6]
```diff
@@ -2180,13 +2180,38 @@ OplMode find_oplmode()
...
+	};
+    switch (find_sbtype()) {
+	case SBT_2: // CMS is optional for SoundBlaster 1, 1.5 and 2
+    case SBT_1: return get_cms();
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2207]: So yeah, just `return sect->Get_bool("cms")`.

File: src/hardware/sblaster.cpp [hunk 7]
```diff
@@ -2196,7 +2221,8 @@ class SBLASTER final {
 	IO_WriteHandleObject write_handlers[0x10] = {};
 
 	static constexpr auto blaster_env_name = "BLASTER";
-	OplMode oplmode = OplMode::None;
+	OplMode oplmode                        = OplMode::None;
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2224]: Minor: Maybe insert a blank line above this, so much blank space looks ridiculous 😄

File: src/hardware/sblaster.cpp [hunk 8]
```diff
@@ -2288,6 +2303,11 @@ class SBLASTER final {
 		} break;
 		}
 
+		cms = find_cms();
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2306]: "find" kinda makes sense for `sbtype` which is an enum, but `is_cms_enabled` is probably a better name here.

**Comment by japsmits** [comment]
Ready for another round. :smile:

File: src/hardware/sblaster.cpp [hunk 9]
```diff
@@ -2169,7 +2169,7 @@ OplMode find_oplmode()
 	} else if (pref == "opl2") {
 		opl_mode = OplMode::Opl2;
 	} else if (pref == "cms") {
-		opl_mode = OplMode::Cms;
+        // Skip for backward compatibility with existing configurations
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2172]: Indentation.

**Comment by johnnovak** [comment]
> Ready for another round. 😄

Changes look great @japsmits, I'll give it a good testing tomorrow!

**Comment by johnnovak** [comment]
Okay @japsmits, the basic use-case works fine, but I think we could refine things a little when changing settings at runtime for less confusing results.

Here are my test results and observations:

## Start with defaults

`sbtype` is `sb16` by default.
CMS is not a valid option with `sb16`, yet the `cms` command returns `true`. ❌

There's a conceptual problem here: `cms` should probably have three settings: `on`, `off`, and `auto` (default). In `auto` mode, CMS should be enabled for `sb1` and `gb`.

`cms` explicitly set to `off` and setting `sbtype = gb` is a weird edge case; in this case, we should force `cms = on` anyway.


## Tests using the default config and changing settings at runtime:

### 1. Setting `sbtype = sb1`

CMS gets auto-enabled, `cms` returns true. ✅
OPL2 is auto-enabled. ✅
`cms = off` disables the CMS. ✅

### 2. Setting `sbtype = gb`

CMS gets auto-enabled, `cms` returns true. ✅
`oplmode = opl2` enables OPL2 (mono) and leaves the CMS alone. ✅
`oplmode = none` disables OPL and leaves the CMS alone. ✅

`cms = off` does nothing (correct ✅) but logs no warning (needs improvement ❌).
We should log a warning when setting `cms = off` when `sbtype` is `gb`.

### 3. Setting `sbtype = sb2`

OPL gets auto-enabled, but not CMS. ✅
`cms = on` turns on the CMS, so it coexists with OPL. ✅
`cms = off` turns off the CMS (and OPL remains enabled). ✅
`oplmode = opl2` enables OPL2 (mono) but not the CMS. ✅

### 4. Setting `sbtype = sbpro1 | sbpro2 | sb16`

`cms = on` does not enable the CMS and does not log an error. ✅
`cms` returns `true`. ✅

This is incorrect: `cms = on` should log a warning and `cms` should be reset to `false`. ❌

Examples of how to set config settings programmatically:

https://github.com/dosbox-staging/dosbox-staging/blob/main/src/gui/render.cpp#L1296-L1300
https://github.com/dosbox-staging/dosbox-staging/blob/main/src/hardware/mixer.cpp#L358-L362

**Comment by japsmits** [comment]
Nice test run, thanks! 👍 

Added the 'auto' setting, which is also force-selected whenever an invalid combination is detected. Since 'auto' is always a valid situation, this prevents warnings when switching between sbtypes.

Also added warnings whenever an invalid combination is detected.

File: src/dosbox.cpp [hunk 10]
```diff
@@ -821,12 +821,21 @@ void DOSBOX_Init()
...
 	               "resampling.");
 
-	const char* oplmodes[] = {"auto", "cms", "opl2", "dualopl2", "opl3", "opl3gold", "none", nullptr};
+	pstring = secprop->Add_string("cms", when_idle, "auto");
+	pstring->Set_values({"on", "off", "auto"});
+	pstring->Set_help(
+	        "Enable CMS emulation for SoundBlaster 1, 1.5 and 2 ('auto' by default).\n"
+	        "  off:  Disable CMS emulation (except when a GameBlaster is selected).\n"
+	        "  on:   Enable CMS emulation on SoundBlaster 1, 1.5 and 2.\n"
+	        "  auto: Enable CMS emulation for SoundBlaster 1 and GameBlaster.");
```
Comment (johnnovak) [inline src/dosbox.cpp:830]: Suggest to go with this:

```
"Enable CMS emulation ('auto' by default).\n"
"  off:   Disable CMS emulation (except when the Game Blaster is selected).\n"
"  on:    Enable CMS emulation on Sound Blaster 1 and 2.\n"
"  auto:  Auto-enable CMS emulation for Sound Blaster 1 and Game Blaster."
```

Main points:

- We don't emulate the Sound Blaster 1.5
- Sound Blaster & Game Blaster (two words)

I will refine the description of `oplmode` shortly myself as it has a lot to be desired 😄

File: src/hardware/sblaster.cpp [hunk 11]
```diff
@@ -2180,13 +2180,61 @@ OplMode find_oplmode()
...
+	const auto sbtype      = find_sbtype();
+	const bool cms_enabled = [sect, sbtype]() {
+		// Backward compatibility with existing configurations
+		if (sect->Get_string("oplmode") == "cms") {
+			LOG_WARNING("%s: The 'cms' setting for 'oplmode' is deprecated; use 'cms = on' instead.",
+			            CardType());
+			return true;
+		} else {
+			const auto cms_str = sect->Get_string("cms");
+			const auto cms_enabled_opt = parse_bool_setting(cms_str);
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2204]: Nice, glad you found it 🏅

File: src/hardware/sblaster.cpp [hunk 12]
```diff
@@ -2180,13 +2180,61 @@ OplMode find_oplmode()
...
+		return true; // GameBlaster is CMS
+	default: 
+		if (cms_enabled) {
+			LOG_WARNING("%s: The 'cms' setting is 'on' but not supported on this card. CMS is forced to 'auto'.",
+			            CardType());
+			auto* sect_updater = static_cast<Section_prop*>(control->GetSection("sblaster"));
+			sect_updater->Get_prop("cms")->SetValue("auto");
+		}
+		return false;
+	}
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2233]: Very nice minimal solution for auto-overriding and syncing the `cms` setting. Works perfectly! 🎉

File: src/hardware/sblaster.cpp [hunk 13]
```diff
@@ -2180,13 +2180,61 @@ OplMode find_oplmode()
...
+		if (!cms_enabled) {
+			LOG_WARNING("%s: The 'cms' setting is 'off', but is forced to 'auto'.",
+			            CardType());
+			auto* sect_updater = static_cast<Section_prop*>(control->GetSection("sblaster"));
+			sect_updater->Get_prop("cms")->SetValue("auto");
+		}
+		return true; // GameBlaster is CMS
+	default: 
+		if (cms_enabled) {
+			LOG_WARNING("%s: The 'cms' setting is 'on' but not supported on this card. CMS is forced to 'auto'.",
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2227]: Minor: `...not supported on this card; forcing 'auto'.` for consistency with other log messages.

File: src/hardware/sblaster.cpp [hunk 14]
```diff
@@ -2180,13 +2180,61 @@ OplMode find_oplmode()
...
+			return false;
+		}
+	}();
+
+	switch (sbtype) {
+	case SBT_2: // CMS is optional for SoundBlaster 1, 1.5 and 2
+	case SBT_1: return cms_enabled;
+	case SBT_GB:
+		if (!cms_enabled) {
+			LOG_WARNING("%s: The 'cms' setting is 'off', but is forced to 'auto'.",
```
Comment (johnnovak) [inline src/hardware/sblaster.cpp:2219]: We have the `GB: ` prefix when this is logged, but this would be a bit clearer:

`...but is forced to 'auto' on the Game Blaster.`

**Review by johnnovak — APPROVED** [review]
This works perfectly now @japsmits, thanks! I really like your simple and elegant solution to deal with the overriding and syncing of the `cms` setting in the various scenarios.

Made a few further comments about the wording of various messages but approving it, very nice work!

**Comment by johnnovak** [comment]
I've invited you as a project collaborator @japsmits. That should make working together easier, should you want to keep your nice stream of PRs going 😄

File: src/dosbox.cpp (hunk 10 above)
Comment (japsmits) [reply src/dosbox.cpp]: Checked that there are no more references to Sound Blaster 1.5. Also corrected the naming of Sound Blaster and Game Blaster in other places (comments only).

**Comment by japsmits** [comment]
Yeah, native speakers always create the better texts. I wonder why... :wink:

Updated all texts, so I hope we're good to go.

**Comment by johnnovak** [comment]
@Python-Exoproject Might be relevant to your interests 😎 Please read the top description.
